esxiSsh.py

```diff
@@ -0,0 +1,288 @@
+        # destVm = string, name of destination VM
+        #    Limitations:
+        #         Source VM must exist on one datastore
+        #         Source VM name must be unique
+        #         esxiServer must not be a vCenter server
+        #         VM cannot contain multiple disks
+
+        if type(srcVm) != str or type(destVm) != str:
+            fatalError("Source and destination VMs must be the VM names as strings.")
+
+        path, srcVmdk = self.findVmdkPath(srcVm)
+
+        session = self.loginToEsx()
+
+        # Make destination directory
+        session.sendline('cd ' + path)
+        session.sendline('mkdir ' + destVm)
+        session.sendline('ls ' + path + '/' + destVm)
+
+        # Copy non-VMDK files from source VM to destination VM
+        session.sendline('find "' + path + '/' + srcVm + '" -maxdepth 1 -type f | grep -v ".vmdk"' + \
+                         ' | while read file; do cp "$file" "' + path + '/' + destVm + '"; done')
+
+        # Copy VMDK files from source VM to destination VM
+        destVmdk = srcVmdk.split('/')[-1]
+        if thinProvision:
+            session.sendline('vmkfstools -i "' + path + '/' + srcVmdk + '" -d thin "' \
+                             + path + '/' + destVm + '/' + destVmdk + '"')
+        else:
+            session.sendline('vmkfstools -i "' + path + '/' + srcVmdk + '" -d zeroedthick "' \
+                             + path + '/' + destVm + '/' + destVmdk + '"')
+
+        # Wait for the VMDK copying to complete, and check for errors
+        while True:
+            i = session.expect(["Clone: 100% done.","Failed to clone disk","Failed to lock the file"], timeout=60*30)
+            if i == 0:
+                break
+            elif i == 1:
+                try:
+                    session.expect("The file already exists", timeout=1)
+                    self.fatalError("The VMDK already exists.  Pick a different destination VM name, or clean up your datastore first.")
+                    return False
+                except pexpect.TIMEOUT:
+                    self.fatalError("An unknown error occured copying the VMDK.  Here, have a shell:")
+                    session.interact()
+            elif i == 2:
+                self.fatalError("Unable to lock the VMDK file.  The VM must be powered off.")
+                return False
+
+        # One last thing, register the new VM in the ESXi inventory
+        session.sendline('vim-cmd solo/registervm ' + path + '/' + destVm + '/' + srcVm + '.vmx ' \
+                         + destVm)
+        try:
+            i = session.expect(['^[0-9]+$'], timeout=30)
+            if i == 0:
+                print session.before
+                print session.after
+            else:
+                print "???"
+            return True
+        except pexpect.TIMEOUT:
+            return False
+
+        session.sendline("exit")
+        session.expect(["Connection to .* closed."],timeout=10)
```

Comment:
This session exit is dead code, should this be moved inside the `except` or always called?

Note: in testing this method in an interactive session I got a `False` result on a successful clone.